docs/features/string-literals-data-section.md

```diff
@@ -0,0 +1,201 @@
+
+```cs
+IFieldReference ITokenDeferral.GetFieldForDataString(ImmutableArray<byte> data, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+```
+
+Similar to `ITokenDeferral.GetFieldForData`, this calls into `PrivateImplementationDetails`
+which is responsible for synthesizing the classes and helpers in a thread-safe and deterministic manner.
+
+There are implementations of various `Cci` interfaces which represent the metadata,
+rooted in a `DataStringHolder` which is the `INamespaceTypeDefinition` for the `<S>` class.
+
+These are implemented in `Microsoft.CodeAnalysis`.
+The alternative (used by inline arrays, for example)
+would be to implement the `Symbol`s in the language-specific projects (and then use the standard `Cci` adapters),
+but that seems to require similar amount of implemented abstract properties/methods of the `Symbol`
+as the implementations of `Cci` interfaces require.
+But implementing `Cci` directly allows us to reuse the same implementation for VB if needed in the future.
+
+## Future work
+
+### Edit and Continue
+
+Hot reload currently does not support `.data` field replacement.
+That can be implemented in the future.
+
+### AOT
+
+Ahead-of-time compilation tools would need to be updated to recognize this new pattern of string literal use in place of `ldstr`.
+
+### Ref assemblies
+
+Need to confirm that no additional work is needed for ref assemblies.
+
+### Configuration/emit alternatives
+
+Instead of one global feature flag, the emit strategy could be controlled using compiler-recognized attributes (applicable to assemblies or classes).
+Furthermore, we could emit more than one string per one class. That could be configurable as well.
+
+### Automatic threshold
+
+The threshold could be determined automatically with some objective, for example,
+use the utf8 encoding emit strategy for the lowest number of string literals necessary to avoid overflowing the UserString heap.
+
+The set of string literals is not know up front in the compiler, it is discovered lazily (and in parallel) by the emit layer.
+However, we could continue emitting `ldstr` instructions and fix them up in a separate phase after we have seen all the string literals
+(and hence can determine the automatic threshold).
+The `ldstr` and `ldsfld` instructions have the same size, so fixup would be a straightforward replace.
+This fixup phase already exists in the compiler in `MetadataWriter.WriteInstructions`
+(it's also how the string literals are emitted into the UserString heap).
+It is called from `SerializeMethodBodies` which precedes `PopulateSystemTables` call,
+hence synthesizing the utf8 string classes in the former should be possible and they would be emitted in the latter.
+
+### Statistics
+
+The compiler could emit an info diagnostic with useful statistics for customers to determine what threshold to set.
+For example, a histogram of string literal lengths encountered while emitting of the program.
+
+## Alternatives
```

Comment:
It might be worth adding another alternative. We might merge all UTF8 byte sequences for affected strings into a single blob - a virtual User-String heap. The sequences can be deduped as we dedupe entries in User-String heap. We even can attempt to do some vary basic compression. For example, if I remember correctly, in the name heap that we generate (contains type and member names, etc.), names with shared content can overlap. We would generate a single ```'__StaticArrayInitTypeSize=*'``` structure for the entire virtual heap blob and add a single RVA field to ```'<PrivateImplementationDetails>'``` that points to the blob. At runtime, we would do an offset to where the required data reside in the blob and decode the required length from Utf-8 to Unicode.